config/routes.rb

```diff
@@ -23,11 +25,15 @@
     # Newer implementation: will supersede the above.
     get "users/two_factor_authentication/setup", to: "users/two_factor_authentication/setup#show"
     put "users/two_factor_authentication/setup", to: "users/two_factor_authentication/setup#update"
-    get "users/two_factor_authentication/email", to: "users/two_factor_authentication/email#show"
-    post "users/two_factor_authentication/email", to: "users/two_factor_authentication/email#update"
+
+    get "users/two_factor_authentication/email/resend", to: "users/two_factor_authentication/email#resend"
+    get "users/two_factor_authentication/email/confirm", to: "users/two_factor_authentication/email#confirm"
+    get "users/two_factor_authentication/app/totp_with_qr", to: "users/two_factor_authentication/app#totp_with_qr"
+    put "users/two_factor_authentication/email", to: "users/two_factor_authentication/email#update"
 
     get "users/:id/two_factor_authentication/edit", to: "users/two_factor_authentication#edit"
     delete "users/:id/two_factor_authentication", to: "users/two_factor_authentication#destroy"
+    get "users/two_factor_authentication/auth/:code", to: "users/two_factor_authentication#update", as: "users_two_factor_authentication_direct_otp"
```

Comment:
Since this is for email OTP, IMHO the path would be clearer if it had the email resource in it (e.g. `..two_factor_authentication/email`), in order to avoid any confusion and since we already have the `..two_factor_authentication/email` resource.

Also, my preference would be to have the code provided as a query string, rather than being a separate entity (e.g. `two_factor_authentication/email?action=auth&code=1234`).